drydock_provisioner/drivers/oob/pyghmi_driver/__init__.py

```diff
@@ -172,115 +172,133 @@ def execute_task(self):
+            worked = False
+            
+            self.logger.debug("Setting bootdev to PXE for %s" % self.node.name)
+            self.exec_ipmi_command(Command.set_bootdev, 'pxe')
+                
             time.sleep(3)
 
-            bootdev = ipmi_session.get_bootdev()
-
+            bootdev = self.exec_ipmi_command(Command.get_bootdev)
+                
             if bootdev.get('bootdev', '') == 'network':
+                self.logger.debug("%s reports bootdev of network" % self.node.name)
                 self.orchestrator.task_field_update(self.task.get_id(),
                     result=hd_fields.ActionResult.Success,
                     status=hd_fields.TaskStatus.Complete)
+                return
             else:
-                self.orchestrator.task_field_update(self.task.get_id(),
+                self.logger.warning("%s reports bootdev of %s" % (ipmi_address, bootdev.get('bootdev', None)))
+                worked = False
+                
+            self.logger.error("Giving up on IPMI command to %s after 3 attempts" % self.node.name)
+            self.orchestrator.task_field_update(self.task.get_id(),
                     result=hd_fields.ActionResult.Failure,
                     status=hd_fields.TaskStatus.Complete)
             return
         elif task_action == hd_fields.OrchestratorAction.PowerOffNode:
-            ipmi_session.set_power('off')
+            worked = False
+
+            self.logger.debug("Sending set_power = off command to %s" % self.node.name)
+            self.exec_ipmi_command(Command.set_power, 'off')
 
             i = 18
 
             while i > 0:
-                power_state = ipmi_session.get_power()
+                self.logger.debug("Polling powerstate waiting for success.")
+                power_state = self.exec_ipmi_command(Command.get_power)
                 if power_state.get('powerstate', '') == 'off':
+                    self.logger.debug("Node reports powerstate of off")
+                    worked = True
                     break
                 time.sleep(10)
                 i = i - 1
 
-            if power_state.get('powerstate', '') == 'off':
+            if worked:
                 self.orchestrator.task_field_update(self.task.get_id(),
                     result=hd_fields.ActionResult.Success,
                     status=hd_fields.TaskStatus.Complete)
             else:
+                self.logger.error("Giving up on IPMI command to %s" % self.node.name)
                 self.orchestrator.task_field_update(self.task.get_id(),
                     result=hd_fields.ActionResult.Failure,
                     status=hd_fields.TaskStatus.Complete)
             return
         elif task_action == hd_fields.OrchestratorAction.PowerOnNode:
-            ipmi_session.set_power('on')
+            worked = False
+
+            self.logger.debug("Sending set_power = off command to %s" % self.node.name)
+            self.exec_ipmi_command(Command.set_power, 'off')
```

Comment:
Agreed, major refactoring needed to functionalize all this now that the working logic is known.